torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,795 @@
+
+        shape = node.meta["val"].shape
+
+        # for reduction op that reduces over the sharded batch dim
+        # we don't generate partial, but rather, we generate shard
+        # This is because the intention of data parallel is to never
+        # do full reduction across batch dimension, it would still
+        # keep the reduction activation sharded.
+        reduction_over_batch = False
+        reduction_ops = [aten.sum.default, aten.mean.default]
+        if node.target in reduction_ops and len(shape) == 0:
+            operand = node.all_input_nodes[0]
+            if operand in self.batch_dim_map:
+                operand_batch_dim = self.get_batch_dim(operand)
+                if operand_batch_dim == 0:
+                    reduction_over_batch = True
+                    self.set_batch_dim(node, operand_batch_dim)
+
+            else:
+                raise RuntimeError(f"batch dim analysis failed on node: {node}!")
+        elif input_full_reduction:
+            # the first consumer node that consumes the full reduction
+            operand = node.all_input_nodes[0]
+            assert (
+                operand in self.batch_dim_map
+            ), "input have full reduction but not in dim map!"
+            self.set_batch_dim(node, self.get_batch_dim(operand))
+        else:
+            for i, dim_size in enumerate(shape):
+                if dim_size == self.batch_dim_size:
+                    self.set_batch_dim(node, i)
+
+        node_batch_dim = self.get_batch_dim(node)
+        batch_dim_shard_spec = DTensorSpec(
+            mesh=mesh, placements=[Shard(node_batch_dim)]
+        )
+        if reduction_over_batch or input_full_reduction:
+            batch_dim_shard_spec.from_local = True  # type: ignore[attr-defined]
+
+        return batch_dim_shard_spec, reduction_over_batch
+
+
+def _gen_shard_strategy(
+    mesh: DeviceMesh, shard_dim: int, input_specs: Optional[List[DTensorSpec]] = None
+) -> PlacementStrategy:
+    """
+    util function to generate a shard strategy on shard_dim
+    """
+    return PlacementStrategy(
+        output_spec=DTensorSpec(mesh=mesh, placements=[Shard(shard_dim)]),
+        input_specs=input_specs,
+    )
+
+
+def _gen_replicate_strategy(
+    mesh: DeviceMesh, input_specs: Optional[List[DTensorSpec]] = None
+) -> PlacementStrategy:
+    """
+    util function to generate a replicate strategy
+    """
+    return PlacementStrategy(
+        output_spec=DTensorSpec(mesh=mesh, placements=[Replicate()]),
+        input_specs=input_specs,
+    )
+
+
+def _gen_partial_strategy(mesh: DeviceMesh) -> PlacementStrategy:
+    """
+    util function to generate a partial strategy
+    """
+    return PlacementStrategy(
+        output_spec=DTensorSpec(mesh=mesh, placements=[_Partial()]),
+    )
+
+
+def build_data_parallel_strategies(
+    train_step_graph: GraphModule,
+    num_params: int,
+    num_states: int,
+    mesh: DeviceMesh,
+    batch_dim: int = 0,
```

Comment:
What exactly does this mean? Is it assumed that all inputs are batched at this dim?

Reply:
Yeah it's a implicit assumption that the input is a mini-batch that is sharded on batch dimension 0 (i.e. a full batch input is sharded on dim 0 to get mini-batch input on each rank), it's being used to initialize the `BatchDimAnalzyer`. This implicit assumption might be too strong, so I pass this as a parameter and plan to expose this to the `DataParallelMode` directly so that it could support cases like batch_dim is not 0-dim.